// less-brunch/lib/index.js
// Generated by CoffeeScript 1.3.3
(function() {
  var LESSCompiler, less, sysPath,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  less = require('less');

  sysPath = require('path');

  module.exports = LESSCompiler = (function() {

    LESSCompiler.prototype.brunchPlugin = true;

    LESSCompiler.prototype.type = 'stylesheet';

    LESSCompiler.prototype.extension = 'less';

    LESSCompiler.prototype._dependencyRegExp = /@import ['"](.*)['"]/g;

    function LESSCompiler(config) {
      this.config = config;
      this.getDependencies = __bind(this.getDependencies, this);

      null;
    }

    LESSCompiler.prototype.compile = function(data, path, callback) {
      var parser,
        _this = this;
      parser = new less.Parser({
        paths: [this.config.paths.root, sysPath.dirname(path)],
        filename: path
      });
      return parser.parse(data, function(error, tree) {
        var err, result;
        if (error != null) {
          return callback(error.message);
        }
        result = null;
        err = null;
        try {
          result = tree.toCSS();
        } catch (ex) {
          err = "" + ex.type + "Error:" + ex.message;
          if (ex.filename) {
            err += " in '" + ex.filename + ":" + ex.line + ":" + ex.column + "'";
          }
        }
        return callback(err, result);
      });
    };

    LESSCompiler.prototype.getDependencies = function(data, path, callback) {
      var dependencies, parent, paths,
        _this = this;
      paths = data.match(this._dependencyRegExp) || [];
      parent = sysPath.dirname(path);
      dependencies = paths.map(function(path) {
        var res;
        res = _this._dependencyRegExp.exec(path);
        _this._dependencyRegExp.lastIndex = 0;
        return (res || [])[1];
      }).filter(function(path) {
        return !!path;
      }).map(function(path) {
        if (sysPath.extname(path) !== ("." + _this.extension)) {
          return path + ("." + _this.extension);
        } else {
          return path;
        }
      }).map(sysPath.join.bind(null, parent));
      return process.nextTick(function() {
        return callback(null, dependencies);
      });
    };

    return LESSCompiler;

  })();

}).call(this);
